### ISSUE TITLE
Future ideas

### ISSUE THREAD

**Issue by tonsky** [tonsky/FiraCode#36]
Markdown: `##` `###` `####`
URL: `://` `:///`
Haskell: `<*` `<+` `<$` `$>` `*>` `+>`

**Comment by tonsky** [comment]
`{{` `}}` (templates)

**Comment by tonsky** [comment]
`{-#` `#-}` (Haskell pragmas)
`<->` ?

**Comment by tonsky** [comment]
Probably return standard `fi` and `fl` but keep them double-width

**Comment by tonsky** [comment]
`!!!` is pretty common

**Comment by ltlollo** [comment]
Hi, yesterday i wrote a script to collect common expressions on various projects scattered around my HD; these are the missing operators that I use the most:

* ```()```
* ```+= (from +=, -=, /=, ...)```
* ```[] (from [], [&], [=])```
* ```{}```

Also I made a repo that gathers expressions and operators across some big projects, https://github.com/ltlollo/S. I hope it' useful.

**Comment by tonsky** [comment]
Thanks!
I considered these combinations, but didn’t end up with any meaningful
visual ideas for them. Maybe you have ideas?

**Comment by ltlollo** [comment]
+= : sould probably just be + closer to = and aligned, as the other operation on self
[], {}, [=] and [&] could have could have end point joined together, except maybe for [], that's also list in haskell

as a demo of my amazing gimp skills: [example](http://imgur.com/1p2ODHB)

**Comment by tonsky** [comment]
What’s [=] and [&] by the way?

**Comment by ltlollo** [comment]
they are lambda captures in c++, http://en.cppreference.com/w/cpp/language/lambda
using [] in a lambda makes it a normal function
```auto fun = [](a) -> b { ... };``` is the same as ```b fun(a){...}```
while with [&] and [=] means you can access variables declared outside of the lambda either by reference [&] or by copy [=].
ex: ```int i = 0; auto fun = [=](...){ return i; };```

**Comment by ltlollo** [comment]
they are lists of captured parameters, but [&]/[=] could be read as box the environment by reference/value

**Comment by tonsky** [comment]
> The only other ligature I’ve found that would be useful for R is %>%, which is used for piping. Also, I believe ## is used for introducing comments.

**Comment by tonsky** [comment]
Clojure’s :>> (from condp)

**Comment by tonsky** [comment]
Scala’s `???` (http://alvinalexander.com/scala/what-does-three-question-marks-in-scala-mean)

**Comment by HalosGhost** [comment]
I'd like to throw in support for all the rest of the assignment operators in C (along with a couple others):

``*=``, ``+=``, ``-=``, ``&=``, ``&&=``, ``%=`` and ``()``.

I would advocate for ``()`` to be similarly stylized to ``<>``.

**Comment by tonsky** [comment]
+>

**Comment by johnmuhl** [comment]
Javascript `!==`

**Comment by gusaiani** [comment]
+1 to !== in js

**Comment by tonsky** [comment]
Hey, it's already there guys

-----Original Message-----

**Comment by gusaiani** [comment]
@tonsky Thanks for pointing it out. 
I don’t know. It may just be Atom.
I'm using 1.2, which does render a lot of glyphs perfectly.
I reinstalled FiraCode .6 but still no `!==` in Atom, or `|=`.
Anyway...seems like it’s not the font's problem. Thanks again. It’s so great to use your ligatures.

**Comment by tonsky** [comment]
Yep. It’s probably tokenization issue. It can’t apply ligature if `!=` and
second `=` are in different spans

**Comment by johnmuhl** [comment]
@tonsky thanks for the tip. Turns out it is a problem with Atom's javascript highlighter. Using the `language-javascript-jsx` package works much better.

**Comment by gusaiani** [comment]
Thanks @johnmuhl. 
Got it to work.
Installed `language-javascript-jsx` and `file-types` packages.
This last one allowed me to set a grammar(?) package to all files of a certain extension, etc.
Then in atom's `confg.cson`, I added:
```
"file-types":
  "source.js": "language-javascript-jsx"
```
and that did it.

Hope this helps someone and thanks again.

**Comment by tonsky** [comment]
`;;;`

**Comment by weavejester** [comment]
A ligature for Clojure's `.-` syntax for attributes might be useful. The ligature could look like `·-`.

**Comment by pmros** [comment]
~>

**Comment by ianp** [comment]
For Scala: `|@|`, `|+|`, and `\/` would be good to have, also `~>` but that's already been mentioned a few times!

Also, is there a way to contribute without having to purchase a copy of [Glyphs](https://glyphsapp.com)?

**Comment by 13h3r** [comment]
Also there are `<:`, `>:` and `<%` in scala

**Comment by tonsky** [comment]
I wonder if Scala contributes more to ligatures than Haskell. Perl
definitely hold third place

**Comment by tomByrer** [comment]
JS's lodash/underscore/ERB use `<% %>` for templates.
https://lodash.com/docs#template

**Comment by honnibal** [comment]
I'd love to see `lambda` as λ, particularly for Python.

**Comment by tonsky** [comment]
@honnibal this would change indentation too much, which is against FC principles

**Comment by emilkaae** [comment]
I love almost all aspects of Fira Code; I think it is really readable and uniform for writing code. I have a problem with the ligatures for #, ##, ### etc. for markdown: I cannot decide the heading level at a glance (like I usually can), but have to either count the dashes and subtract one (wtf?) or sieve through with the cursor. I'd _really_ appreciate if the markdown features were toned down.

This particular issue has made me try Monoid for markdown editing, although it takes a little getting used to as it is much more narrow.

**Comment by tonsky** [comment]
@emilkaae I know what are you talking about. I’ll figure something out #287

**Comment by danbeck** [comment]
I would love to have a ligature for the "return" keyword. For example: ↳

**Comment by PEZ** [comment]
Clojure: Lambda ligature for `(fn`, something like `(λ`

It would maybe call for something for `#(` as well? `ƛ(`?

**Comment by tonsky** [comment]
Unfortunately `(λ` wouldn’t work, as it has to be 2-char wide (same width as `fn`)

**Comment by PEZ** [comment]
It'll have to be a Phat Lambda™ then :smile:

**Comment by tonsky** [comment]
`(\` might conflict with regexes, I think.

**Comment by RubenVerg** [comment]
> I wonder if Scala contributes more to ligatures than Haskell. Perl
> definitely hold third place

challenge: someone make a language which has 1000000 ligatures for basic stuff, like addition is `>=+=<@.` or something (that means,
- `>` get the value to the left
- `=` pipe it to `+`
- `+` add, when I have 2 args
- `=` pipe
- `<` get value to the right
- `@` place it
- `.` here
)

**Comment by akuklev** [comment]
For mathematical texts (mainly in theorem provers) the following characters would be desirable:
* Uppercase and lowercase Greek letters including several letter variants (\varbeta, \varepsilon, \vartheta, \varkappa, \varpi, \varrho, \varsigma, \varphi, and inverted delda and omega):  Α α, Β β/ϐ, Γ γ, Δ δ, Ε ε/ϵ, Ζ ζ, Η η, Θ θ/ϑ, Ι ι, Κ κ/ϰ, Λ λ, Μ μ, Ν ν, Ξ ξ, Ο ο, Π π/ϖ, Ρ ρ/ϱ, Σ σ/ς, Τ τ, Υ υ, Φ φ/ϕ, Χ χ, Ψ ψ, and Ω ω, and ∇, ℧.
* Quantifiers, large sum and large product characters: ∀, ∃, ∄, ∏, ∐, ∑, ⋂, ⋃, 
* Several letterlike and blackboard bold characters: 𝔹, ℂ, 𝔻, 𝔼, 𝔽, ℍ, ℕ, ℙ, ℚ, ℝ, and ℤ; ℑ, ℜ, ℵ, ℶ, ℘, ℬ, ∅, ∂, ∞.
* Logical symbols: ¬, ⅋, ∧, ∨, ⊼, ⊻, ⊤, ⊥, ⊢, ⊣, ⊨, ⫤, ⊸
* Brackets: ‹›, ⌈⌉, ⌊⌋,⁅⁆, ⁽⁾,₍₎,〈〉,❬❭, ❲❳, ⟦⟧, ⧘⧙.
* Basic operators: ⋅, ×, ⋉, ⋊, ∘, ∗, ⋆, ⊙, ⊕, ⊖, ⊗, ⊛, ⊚.

**Comment by timwisefmg** [comment]
> I'd love to see `lambda` as λ, particularly for Python.
> 
> I'd be in favour of the other greek letters as well, although I figure they'd be more controversial.

Would love to see this for Haskell ('\' as 'λ')
